lib/resque/scheduler.rb

```diff
@@ -175,19 +175,19 @@ def env_matches?(configured_env)
       # Handles queueing delayed items
       # at_time - Time to start scheduling items (default: now).
       def handle_delayed_items(at_time = nil)
-        timestamp = Resque.next_delayed_timestamp(at_time)
-        if timestamp
-          procline 'Processing Delayed Items'
-          until timestamp.nil?
-            enqueue_delayed_items_for_timestamp(timestamp)
-            timestamp = Resque.next_delayed_timestamp(at_time)
+        at_time ||= Time.now
+        item = Resque.next_delayed_item(before: at_time)
+        procline 'Processing Delayed Items'
+
+        until item.nil?
+          handle_shutdown do
```

Comment:
That's no different than the previous code. `Resque.next_item_for_timestamp(timestamp)` can lead to lost jobs as well. There is a tradeoff... I think we either risk lost jobs our double-running jobs in case of failure.

Reply:
I think we should be more careful here.

I see a difference in that the previous code would lose the job in exceptional circumstances (SIGKILL basically is the only case I can think of since other signals are trapped).

Whereas, in your changes, you pop the item off the queue (L186) and _immediately_ exit (L183) if any of the signals have been set:

```
      def handle_signals
          ...
          case sig
          when 'INT', 'TERM', 'QUIT' then shutdown
      end
```

```
      def shutdown
        return if @shutdown
        @shutdown = true
        log!('Shutting down')
        @th.raise Interrupt if @sleeping
      end
```


```
      def handle_shutdown
        exit if @shutdown
        yield
        exit if @shutdown
      end
````

Reply:
What do you suggest then? It's a tricky situation.

Reply:
anything that runs the enqueueing code before `exit if @shutdown` should be fine.

Could push the `Resque.next_delayed_item` call to `enqueue_delayed_item`

Could also reverse the order of the calls in your loop

There's probably other ways

Reply:
Yea I think converting to a `do while` format should be sufficient.